crates/re_data_store/src/entity_tree.rs

```diff
@@ -301,45 +360,82 @@ impl EntityTree {
     /// Purge all times before the cutoff, or in the given set
     pub fn purge(
         &mut self,
-        cutoff_times: &BTreeMap<Timeline, TimeInt>,
-        drop_row_ids: &ahash::HashSet<RowId>,
+        deleted: &re_arrow_store::Deleted,
+        deleted_by_us_and_children: &mut ActuallyDeleted,
     ) {
         let Self {
-            path: _,
+            path,
             children,
             prefix_times,
-            num_timeless_messages: _,
+            num_timeless_messages,
             nonrecursive_clears,
             recursive_clears,
-            components: fields,
+            components,
         } = self;
 
+        if let Some(decrement) = deleted.timeless.get(&path.hash()) {
+            *num_timeless_messages = num_timeless_messages.saturating_sub(*decrement as _);
+            deleted_by_us_and_children.timeless += decrement;
+        }
+
         {
             re_tracing::profile_scope!("prefix_times");
-            for (timeline, histogram) in &mut prefix_times.0 {
-                if let Some(cutoff_time) = cutoff_times.get(timeline) {
-                    histogram.remove(..cutoff_time.as_i64());
-                }
-            }
+            prefix_times.purge(path.hash(), deleted, deleted_by_us_and_children);
         }
         {
             re_tracing::profile_scope!("nonrecursive_clears");
-            nonrecursive_clears.retain(|row_id, _| !drop_row_ids.contains(row_id));
+            nonrecursive_clears.retain(|row_id, _| !deleted.row_ids.contains(row_id));
         }
         {
             re_tracing::profile_scope!("recursive_clears");
-            recursive_clears.retain(|row_id, _| !drop_row_ids.contains(row_id));
+            recursive_clears.retain(|row_id, _| !deleted.row_ids.contains(row_id));
         }
 
         {
-            re_tracing::profile_scope!("fields");
-            for columns in fields.values_mut() {
-                columns.purge(cutoff_times);
+            re_tracing::profile_scope!("ComponentStats");
+            for stats in components.values_mut() {
+                let ComponentStats {
+                    times,
+                    num_timeless_messages,
+                } = stats;
+
+                times.purge(path.hash(), deleted, deleted_by_us_and_children);
```

Comment:
As mentioned above, this purge should only happen if this component was actually cleared, which isn't guaranteed.